### ISSUE TITLE
How to show alteration type?

### ISSUE THREAD

**Issue by yiyanliao** [NKI-CCB/DISCOVER#30]
As the documentary said, the R/Python package can only perform **Groupwise test** and finally get a plot without alteration type information, not like the ones in the original paper. I tried to reproduce repo [discover-notebooks](https://github.com/scanisius/discover-notebooks), but it turned out that the data cannot be downloaded anymore, therefore, I have no clue where alteration information in the final plot of [MSigDb group tests](https://github.com/scanisius/discover-notebooks/blob/master/notebooks/MSigDb%20group%20tests.ipynb) comes from.

**Comment by yiyanliao** [comment]
I noticed that there is an explanation in #17, but it's not the case with the final plot, as I said before, in the paper. I would appreciate it really much if I can get a solution/further explanation about this confusion.

**Comment by scanisius** [comment]
To perform group-wise tests for mixed alteration types, there are a couple of steps to take:

1. Construct binary alteration matrices for each individual alteration type. For the analyses in the DISCOVER paper, we analysed three alteration types: mutation, copy number gain, and copy number loss. For each of those we constructed a binary alteration matrix with genes in the rows and tumours in the columns, just like you would when performing a DISCOVER analysis for an individual alteration type.
   
   To illustrate the concept, I'll use the example data that comes with the DISCOVER package to create three such matrices. In reality these data correspond to a single alteration type, but for the example let's pretend we can split them in three different alteration types.

```python
import discover
import discover.datasets
import numpy

brca_mut = discover.datasets.load_dataset("brca_mut")

# randomly assign the observed mutations to the three alteration types
# for your own analyses, replace these by the actual alteration matrices
alterations_mut = brca_mut & numpy.random.binomial(1, 0.3, size=brca_mut.shape)
alterations_gain = brca_mut & numpy.random.binomial(1, 0.3, size=brca_mut.shape)
alterations_loss = brca_mut & numpy.random.binomial(1, 0.3, size=brca_mut.shape)
```

   We will be combining these matrices later, so it is helpful to add a suffix to the gene names to distinguish different alterations of the same gene.

```python
alterations_mut = alterations_mut.rename(index=lambda gene: gene + "_mut")
alterations_gain = alterations_gain.rename(index=lambda gene: gene + "_gain")
alterations_loss = alterations_loss.rename(index=lambda gene: gene + "_loss")
```

2. Since each type of alteration will have its own background rates, we fit the DISCOVER background model to each of the alteration matrices individually.

```python
events_mut = discover.DiscoverMatrix(alterations_mut)
events_gain = discover.DiscoverMatrix(alterations_gain)
events_loss = discover.DiscoverMatrix(alterations_loss)
```

3. Combine the background models into a single model.

```python
events = discover.row_stack([events_mut, events_gain, events_loss])
```

4. Finally, to perform a group-wise test with mixed alteration types, select all genes in your gene set of interest and all alteration types.

```python
discover.groupwise_discover_test(events[["TP53_mut", "TP53_loss", "MDM2_mut", "MDM2_gain"]])
```

**Comment by yiyanliao** [comment]
> To perform group-wise tests for mixed alteration types, there are a couple of steps to take:
> 
> 1. Construct binary alteration matrices for each individual alteration type. For the analyses in the DISCOVER paper, we analysed three alteration types: mutation, copy number gain, and copy number loss. For each of those we constructed a binary alteration matrix with genes in the rows and tumours in the columns, just like you would when performing a DISCOVER analysis for an individual alteration type.
>    To illustrate the concept, I'll use the example data that comes with the DISCOVER package to create three such matrices. In reality these data correspond to a single alteration type, but for the example let's pretend we can split them in three different alteration types.
> 
> import discover
> import discover.datasets
> import numpy
> 
> brca_mut = discover.datasets.load_dataset("brca_mut")
> 
> # randomly assign the observed mutations to the three alteration types
> # for your own analyses, replace these by the actual alteration matrices
> alterations_mut = brca_mut & numpy.random.binomial(1, 0.3, size=brca_mut.shape)
> alterations_gain = brca_mut & numpy.random.binomial(1, 0.3, size=brca_mut.shape)
> alterations_loss = brca_mut & numpy.random.binomial(1, 0.3, size=brca_mut.shape)
> We will be combining these matrices later, so it is helpful to add a suffix to the gene names to distinguish different alterations of the same gene.
> 
> alterations_mut = alterations_mut.rename(index=lambda gene: gene + "_mut")
> alterations_gain = alterations_gain.rename(index=lambda gene: gene + "_gain")
> alterations_loss = alterations_loss.rename(index=lambda gene: gene + "_loss")
>  2. Since each type of alteration will have its own background rates, we fit the DISCOVER background model to each of the alteration matrices individually.
> 
> events_mut = discover.DiscoverMatrix(alterations_mut)
> events_gain = discover.DiscoverMatrix(alterations_gain)
> events_loss = discover.DiscoverMatrix(alterations_loss)
>  3. Combine the background models into a single model.
> 
> events = discover.row_stack([events_mut, events_gain, events_loss])
>  4. Finally, to perform a group-wise test with mixed alteration types, select all genes in your gene set of interest and all alteration types.
> 
> discover.groupwise_discover_test(events[["TP53_mut", "TP53_loss", "MDM2_mut", "MDM2_gain"]])

Thanks for your kind reply! I've tried a similar approach, but I found that the **p-values** of the original approach is different from the ones using separate alteration types. For example, my data used to look like this (I used the R package of DISCOVER):
```R
> final_matrix[1:5, 1:5]
         DO10486 DO10631 DO14192 DO14264 DO14343
DPM1           1       1       1       1       1
SCYL3          1       1       0       0       0
C1orf112       0       0       0       0       0
FGR            1       0       0       0       0
CFH            0       0       0       0       0
```

when I perform groupwise test using `genes <- c('TP53','MDM2')`, I got these results:

```R
> groupwise.discover.test(events[genes, ], method="impurity")
Groupwise DISCOVER mutual exclusivity test

method: impurity
P = 0.1626657
> groupwise.discover.test(events[genes, ], method="coverage")
Groupwise DISCOVER mutual exclusivity test

method: coverage
P = 0.3001296
> groupwise.discover.test(events[genes, ], method="exclusivity")
Groupwise DISCOVER mutual exclusivity test

method: exclusivity
P = 0.1472657
```

![Image](https://github.com/user-attachments/assets/39754a6a-0f92-4654-83e8-a0d4f467a1ef)

however, we I transform my original data to the format that contains alteration types, I got these:

```R
> final_matrix[1:5, 1:5] 
                  DO10486 DO10631 DO14192 DO14264 DO14343
DPM1:expr_outlier       1       0       1       0       0
DPM1:cn                 0       1       0       1       1
DPM1:variants           0       0       0       0       0
DPM1:isSplice           0       0       0       0       0
DPM1:fusion             0       0       0       0       0

...

genes <- c('TP53:alt_prom','MDM2:alt_prom',
           'TP53:expr_outlier','MDM2:expr_outlier',
           'TP53:rna_edit','MDM2:rna_edit',
           'TP53:variants','MDM2:variants',
           'TP53:ase_all','MDM2:ase_all',
           'TP53:fusion','MDM2:fusion',
           'TP53:isSplice','MDM2:isSplice',
           'TP53:cn','MDM2:cn')
genes <- genes[genes %in% rownames(final_matrix)]

> groupwise.discover.test(events[genes, ], method="impurity")
Groupwise DISCOVER mutual exclusivity test

method: impurity
P = 0.5404529
> groupwise.discover.test(events[genes, ], method="coverage")
Groupwise DISCOVER mutual exclusivity test

method: coverage
P = 0.393061
> groupwise.discover.test(events[genes, ], method="exclusivity")
Groupwise DISCOVER mutual exclusivity test

method: exclusivity
P = 0.4057534
```

![Image](https://github.com/user-attachments/assets/e0e38368-4887-43ae-a131-5d2d8ef14968)

These two results are not aligned with each other.
Actually I wonder if there's a way to make plots that in original DISCOVER paper, which I can show alteration types in a single gene instead of making them into different "genes", just like this:

![Image](https://github.com/user-attachments/assets/31fc6370-791a-439b-9ba8-e9bb684e330e)

Thanks for your kind reply and I'm looking forward to your further suggestions on this.

**Comment by yiyanliao** [comment]
Finally I wrote the scripts by myself. Here's a brief tutorial.
Let's pretend you have dataset like this:
```R
> head(master)
# A tibble: 6 × 11
  hgnc_symbol ICGC_DONOR_ID Cancer_Type      alt_prom expr_outlier rna_edit variants ase_all fusion isSplice
  <chr>       <chr>         <chr>               <dbl>        <dbl>    <dbl>    <dbl>   <dbl>  <dbl>    <dbl>
1 DPM1        DO10486       ColoRect-AdenoCA        0            1        0        0       0      0        0
2 DPM1        DO10631       ColoRect-AdenoCA        0            0        0        0       0      0        0
3 DPM1        DO14192       Head-SCC                0            1        0        0       0      0        0
4 DPM1        DO14264       Head-SCC                0            0        0        0       0      0        0
5 DPM1        DO14343       Head-SCC                0            0        0        0       0      0        0
6 DPM1        DO15014       Head-SCC                0            1        0        0       0      0        0
```
And these are the genes you want to test with `groupwise.discover.test`:
```R
genes <- c('CDKN2A','PIK3CA','PTEN','AKT3','AKT1','SOS1')
```
Because in my data, there will be more than one alteration for a gene in a single donor, I need to define a priority order and get the final alteration type for a gene in each donor:
```R
priority_order <- c("variants", "fusion", "cn", "ase_all", "alt_prom", "expr_outlier", "rna_edit", "isSplice")

master$top_alt <- apply(master[priority_order], 1, function(x) {
  types <- priority_order[x > 0]
  if(length(types) == 0) return("none") else return(types[1])
})
```
Then you need to extract data for your genes, and define a color palette:
```R
type_matrix <- master %>%
  filter(hgnc_symbol %in% genes) %>%
  select(hgnc_symbol, ICGC_DONOR_ID, top_alt) %>%
  pivot_wider(names_from = ICGC_DONOR_ID, values_from = top_alt, values_fill = "none") %>%
  column_to_rownames("hgnc_symbol") %>%
  as.matrix()

alt_colors <- brewer.pal(min(length(priority_order), 8), "Set2")
names(alt_colors) <- priority_order
alt_colors <- c(none = "lightgray", alt_colors)  
```
Then, modified from the original [R script](https://github.com/NKI-CCB/DISCOVER/blob/master/R/R/plot.R) used in DISCOVER to generate plots, we get these functions:
```R
nested_sort_matrix <- function(alt_matrix, priority_order, genes) {
  blocks <- list(seq_len(ncol(alt_matrix)))
  
  for (gene in genes) {
    new_blocks <- list()
    
    for (block in blocks) {
      types <- alt_matrix[gene, block]
      
      type_df <- data.frame(
        type = c(priority_order, "none"),
        priority = c(seq_along(priority_order), Inf)
      )
      
      block_df <- data.frame(
        sample_idx = block,
        type = types,
        block_pos = seq_along(block)
      ) %>%
        left_join(type_df, by = "type") %>%
        arrange(priority, block_pos)  
      
      type_groups <- rle(as.vector(block_df$type))$values
      group_bounds <- c(0, cumsum(rle(as.vector(block_df$type))$lengths))
      
      for (i in seq_along(type_groups)) {
        start_idx <- group_bounds[i] + 1
        end_idx <- group_bounds[i+1]
        new_blocks[[length(new_blocks)+1]] <- block_df$sample_idx[start_idx:end_idx]
      }
    }
    blocks <- new_blocks
  }
  
  gene_order <- order(rowSums(alt_matrix != "none"), decreasing = TRUE)
  final_order <- unlist(blocks)
  return(alt_matrix[gene_order, final_order, drop = FALSE])
}

plot.alt_matrix <- function(alt_matrix, color_mapping, sort = TRUE) {
  if(sort) {
    sorted_alt_matrix <- nested_sort_matrix(alt_matrix, priority_order, genes)
  } else {
    sorted_alt_matrix <- alt_matrix
  }
  
  ngenes <- nrow(sorted_alt_matrix)
  nsamples <- ncol(sorted_alt_matrix)
  coverage <- sum(colSums(sorted_alt_matrix != "none") > 0)
  
  numOfOncos <- ngenes * nsamples
  oncoCords <- matrix(rep(0, numOfOncos * 5), nrow = numOfOncos)
  colnames(oncoCords) <- c("xleft", "ybottom", "xright", "ytop", "type")
  
  xpadding <- 0.01
  ypadding <- 0.01
  cnt <- 1
  
  for(i in 1:ngenes) {
    for(j in 1:nsamples) {
      xleft <- j - 1 + xpadding
      ybottom <- ((ngenes - i + 1) - 1) + ypadding
      xright <- j - xpadding
      ytop <- (ngenes - i + 1) - ypadding
      type <- as.character(sorted_alt_matrix[i, j])
      
      oncoCords[cnt, ] <- c(xleft, ybottom, xright, ytop, type)
      cnt <- cnt + 1
    }
  }
  
  colors <- color_mapping[oncoCords[, "type"]]
  colors[is.na(colors)] <- "lightgray"
  
  par(pin = c(5.2, 3.2), mar = c(5, 7, 3, 3))
  plot(c(0, nsamples), c(0, ngenes), type = "n",
       main = sprintf("Gene set altered in %.2f%%: %d of %d cases", 
                      coverage/nsamples*100, coverage, nsamples),
       xlab = "Sample Index", ylab = "", yaxt = "n", xaxt = "n")
  
  for(k in 1:nrow(oncoCords)) {
    rect(oncoCords[k, "xleft"], oncoCords[k, "ybottom"], 
         oncoCords[k, "xright"], oncoCords[k, "ytop"],
         col = colors[k], border = NA)
  }
  
  axis(2, at = (ngenes:1) - 0.5, labels = rownames(sorted_alt_matrix), las = 2)
  
  tick_positions <- seq(0.5, nsamples - 0.5, by = 100)
  tick_labels <- seq(0, nsamples, by = 100) 
  axis(1, at = tick_positions, labels = tick_labels, las = 1)
}
```
Then you can plot with color:
```R
plot.alt_matrix(type_matrix, alt_colors, sort = TRUE)
```

<img width="1384" alt="Image" src="https://github.com/user-attachments/assets/f389696a-4dc7-4e93-8f26-773d4c8c9e05" />

If you need a figure legend, use this:

```R
plot.legend <- function(color_mapping) {
  num_colors <- length(color_mapping)
  plot(NULL, xlim = c(0, 1), ylim = c(0, 1), axes = FALSE, xlab = "", ylab = "")
  
  legend_width <- 0.2
  legend_height <- 0.8
  legend_x <- 0.1
  legend_y <- 0.1
  
  cell_width <- legend_width / 1
  cell_height <- legend_height / num_colors
  
  for (i in 1:num_colors) {
    rect(legend_x, legend_y + (num_colors - i) * cell_height,
         legend_x + cell_width, legend_y + (num_colors - i + 1) * cell_height,
         col = color_mapping[i], border = "black")
    text(legend_x + cell_width + 0.05, 
         legend_y + (num_colors - i + 0.5) * cell_height,
         names(color_mapping)[i], adj = 0)
  }
}
```

Then plot:
```R
plot.legend(alt_colors)
```

<img width="190" alt="Image" src="https://github.com/user-attachments/assets/aee478ec-d681-405a-8ce4-b27d958a152c" />

Hope this is helpful. Have fun!
